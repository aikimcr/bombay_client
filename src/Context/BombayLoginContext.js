import { createContext } from "react";

const BombayLoginContext = createContext({
  loggedIn: false,
  showLoginForm: false,
  setLoggedIn: null,
  setShowLogin: null,
});

export default BombayLoginContext;
