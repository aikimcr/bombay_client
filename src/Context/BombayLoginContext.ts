import { createContext, Dispatch } from "react";

interface BombayLoginContextValues {
  loggedIn: boolean;
  setLoggedIn?: Dispatch<boolean>;
}

export const BombayLoginContext = createContext({
  loggedIn: false,
  setLoggedIn: null,
});

export default BombayLoginContext;
