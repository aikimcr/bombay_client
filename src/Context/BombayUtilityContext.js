import { createContext } from 'react';

const BombayUtilityContext = createContext({});

export default BombayUtilityContext;
