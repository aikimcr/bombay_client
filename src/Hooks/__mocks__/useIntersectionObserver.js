import { useEffect, useRef } from "react";

export class mockEntry {
  constructor(target, isIntersecting = false) {
    this.target = target;
    this.isIntersecting = isIntersecting;
    this.isVisible = false;
  }
}

export class mockObserver {
  static observers = [];

  constructor(callback, options) {
    this.callback = callback;
    this.options = options;
    this.entries = [];
    this.constructor.observers.push(this);
  }

  observe(target) {}

  unobserve(target) {}

  disconnect() {}

  _fireIntersect(target, isIntersecting = true) {
    const entry = new mockEntry(target, isIntersecting);
    this.callback([entry], this);
  }
}

export function useIntersectionObserver(topRef, callback, options = {}) {
  const observer = useRef(null);

  useEffect(() => {
    if (topRef.current == null) return;

    if (observer.current == null) {
      observer.current = new mockObserver(callback, {
        ...options,
        root: topRef.current.parentElement,
      });
    }

    return () => {
      observer.current.disconnect();
    };
  }, [observer.current, topRef]);

  return observer;
}

export default useIntersectionObserver;
