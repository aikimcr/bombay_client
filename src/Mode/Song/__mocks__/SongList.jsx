import React from 'react';
import { ComponentMock } from '../../../testHelpers/mocks/ComponentMock';

export const SongList = (props) => {
  return (
    <ComponentMock
      {...props}
      name="SongList"
      className="mock-song-list"
      testId="mock-song-list"
    />
  );
};
