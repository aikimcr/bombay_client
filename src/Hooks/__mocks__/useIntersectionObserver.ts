import { useEffect, useRef, RefObject } from 'react';

import { IntersectionObserverOptions } from '../useIntersectionObserver';
export class mockEntry {
  boundingClientRect: DOMRectReadOnly;
  intersectionRatio: number;
  intersectionRect: DOMRectReadOnly;
  isIntersecting: boolean;
  rootBounds: DOMRectReadOnly;
  target: Element;
  time: number;
  isVisible: boolean;

  constructor(target: Element, isIntersecting = false) {
    this.target = target;
    this.isIntersecting = isIntersecting;
    this.isVisible = false;
  }
}

export class mockObserver {
  root: Element;
  rootMargin: string;
  thresholds: readonly number[];
  callback: IntersectionObserverCallback;
  options: IntersectionObserverOptions;
  entries: mockEntry[];

  static observers: mockObserver[] = [];

  constructor(
    callback: IntersectionObserverCallback,
    options: IntersectionObserverOptions,
  ) {
    this.callback = callback;
    this.options = options;
    this.entries = [];
    mockObserver.observers.push(this);
  }

  observe(_target: Element) {}

  unobserve(_target: Element) {}

  takeRecords() {
    return this.entries;
  }

  disconnect() {}

  _fireIntersect(target: Element, isIntersecting = true) {
    const entry = new mockEntry(target, isIntersecting);
    this.callback([entry], this);
  }
}

export function useIntersectionObserver(
  topRef: RefObject<Element>,
  callback: IntersectionObserverCallback,
  options: IntersectionObserverOptions = {},
) {
  const observer = useRef(null);

  useEffect(() => {
    if (topRef.current == null) {
      return;
    }

    if (observer.current == null) {
      observer.current = new mockObserver(callback, {
        ...options,
        root: topRef.current.parentElement,
      });
    }

    return () => {
      observer.current.disconnect();
    };
  }, [observer.current, topRef]);

  return observer;
}

export default useIntersectionObserver;
