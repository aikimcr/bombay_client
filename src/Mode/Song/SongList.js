import { useEffect, useState, useRef, createRef } from 'react';
import useIntersectionObserver from '../../Hooks/useIntersectionObserver';

import './SongList.scss';

import SongCollection from '../../Model/SongCollection';

import SongListItem from './SongListItem';
import Song from './Song';
import FormModal from '../../Modal/FormModal';

function SongList(props) {
    const topRef = createRef();

    const songCollection = useRef(null);
    const observer = useIntersectionObserver(topRef, (entries, observer) => {
        entries.forEach(entry => {
            if (entry.isIntersecting) {
                setShouldPage(true);
                observer.unobserve(entry.target);
            }

            // I haven't found a case where this is actually needed, but I haven't ruled it out.
            // if (entry.isVisible) {
            // }
        });
    });

    const [showAdd, setShowAdd] = useState(false);
    const [shouldPage, setShouldPage] = useState(false);
    const [loading, setLoading] = useState(true);

    useEffect(() => {
        if (songCollection.current == null) {
            console.count('Init Songlist');
            refreshCollection();
        }

        const intervalHandle = setInterval(refreshCollection, 300000);

        return () => {
            clearInterval(intervalHandle);
        }
    }, []);

    useEffect(() => {
        if (!shouldPage) { return; }
        if (loading) { return; }

        setShouldPage(false);

        if (songCollection.current.hasNextPage()) {
            setLoading(true);

            songCollection.current.fetchNextPage()
                .then(() => {
                    setLoading(false);
                });
        }
    }, [shouldPage, loading]);

    useEffect(() => {
        if (loading) { return; }

        if (songCollection.current && songCollection.current.hasNextPage()) {
            const myElement = topRef.current.querySelector('li:last-child');

            if (myElement) {
                observer.current.observe(myElement);
            }
        }
    }, [loading, observer, topRef]);

    function refreshCollection() {
        setLoading(true);

        songCollection.current = new SongCollection();
        songCollection.current.ready()
            .then(() => {
                setLoading(false);
            });
    }

    async function submitNewSong(songDef) {
        await songCollection.current.save(songDef);
        refreshCollection();
    }

    return (
        <div className="list-component">
            <div className="list-controls">
                <button className="btn" onClick={() => setShowAdd(true)}>New</button>
                <button className="btn" onClick={refreshCollection}>Refresh</button>
                <FormModal
                    title="Add Song"
                    onClose={() => setShowAdd(false)}
                    onSubmit={submitNewSong}
                    open={showAdd}>
                    <Song />
                </FormModal>
            </div>
            <div className="song-list-container" ref={topRef}>
                <ul className="song-list card-list">
                    {songCollection?.current == null ? '' : songCollection.current.map(song => {
                        const key = `song-list-${song.get('id')}`;
                        return <SongListItem className key={key} song={song} />
                    })}
                </ul>
            </div>
        </div>
    );
}

export default SongList;