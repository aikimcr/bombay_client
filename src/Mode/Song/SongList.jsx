import { useState, useContext, createRef } from 'react';

import './SongList.scss';

import BombayLoginContext from '../../Context/BombayLoginContext';

import { SongCollection } from '../../Model/SongCollection';

import SongListItem from './SongListItem';
import { Song } from './Song';
import FormModal from '../../Modal/FormModal';
import { useModelCollection } from '../../Hooks/useModelCollection';
import { ProtectedRoute } from '../../Components';

export const SongList = (props) => {
  const topRef = createRef();

  const loginState = useContext(BombayLoginContext);

  const [songCollection] = useState(new SongCollection({}));
  const { refreshCollection } = useModelCollection({
    initialCollection: songCollection,
    topRef,
  });

  const [showAdd, setShowAdd] = useState(false);

  async function submitNewSong(songDef) {
    await songCollection.save(songDef);
    refreshCollection();
  }

  return (
    <ProtectedRoute>
      <div className="list-component" data-testid="song-list-component">
        <div className="list-controls">
          <button className="btn" onClick={() => setShowAdd(true)}>
            New
          </button>
          <div className="title">Songs</div>
          <button className="btn" onClick={refreshCollection}>
            Refresh
          </button>
          <FormModal
            title="Add Song"
            onClose={() => setShowAdd(false)}
            onSubmit={submitNewSong}
            open={showAdd}
          >
            <Song />
          </FormModal>
        </div>
        <div className="song-list-container list-container" ref={topRef}>
          <ul className="song-list card-list">
            {songCollection == null
              ? ''
              : songCollection.map((song) => {
                  const key = `song-list-${song.id}`;
                  return <SongListItem className key={key} song={song} />;
                })}
          </ul>
        </div>
      </div>
    </ProtectedRoute>
  );
};
