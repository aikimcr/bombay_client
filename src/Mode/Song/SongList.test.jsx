import { useState } from 'react';
import { BrowserRouter } from 'react-router';

import { act, render, screen } from '@testing-library/react';

import {
  mockLoginStatus,
  mockRefreshToken,
  mockLogin,
  mockLogout,
} from '../../Network/testing';

jest.mock('../../Network/Login', () => {
  const originalModule = jest.requireActual('../../Network/Login');

  return {
    __esModule: true,
    ...originalModule,
    loginStatus: mockLoginStatus,
    refreshToken: mockRefreshToken,
    login: mockLogin,
    logout: mockLogout,
  };
});

import { mockModelFetcher, mockUseModelCollection } from '../../Hooks/testing';
jest.mock('../../Hooks/useModelCollection', () => {
  const originalModule = jest.requireActual('../../Hooks/useModelCollection');

  return {
    __esModule: true,
    ...originalModule,
    useModelCollection: mockUseModelCollection,
  };
});

import * as mockObserver from '../../Hooks/useIntersectionObserver';
jest.mock('../../Hooks/useIntersectionObserver');

import { makeModels, makeAModel } from '../../testHelpers/modelTools';

import BombayLoginContext from '../../Context/BombayLoginContext';
import BombayUtilityContext from '../../Context/BombayUtilityContext';

import { SongList } from './SongList.jsx';

jest.useFakeTimers();

function FakeContent(props) {
  const [loginState, setLoginState] = useState({
    loggedIn: true,
    showLoginForm: false,
  });
  const [modeState, setModeState] = useState('song');

  const checkLoginState = async () => {
    setLoginState(true);
  };

  const setMode = async (newMode) => {
    setModeState('song');
  };

  const getBootstrap = () => {
    return {
      keySignatures: ['A', 'B', 'C', 'D', 'E', 'F', 'G'],
    };
  };

  const utilities = {
    checkLoginState,
    setMode,
    getBootstrap,
  };

  return (
    <BrowserRouter basename="/">
      <BombayLoginContext.Provider value={loginState}>
        <BombayUtilityContext.Provider value={utilities}>
          {props.children}
        </BombayUtilityContext.Provider>
      </BombayLoginContext.Provider>
    </BrowserRouter>
  );
}

jest.useFakeTimers();

const testToken =
  'eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJzdWIiOiJGREM4MTEzOCIsInVzZXIiOnsiaWQiOjEsIm5hbWUiOiJhZG1pbiIsImFkbWluIjpmYWxzZX0sImlhdCI6MTY2NTk2NTA5OX0.2vz14X7Tm-oFlyOa7dcAF-5y5ympi_UlWyJNxO4xyS4';

beforeEach(() => {
  // const modalRoot = document.createElement("div");
  // modalRoot.id = "modal-root";
  // document.body.append(modalRoot);

  localStorage.setItem('jwttoken', testToken);
});

afterEach(() => {
  while (mockObserver.mockObserver.observers.length > 0) {
    mockObserver.mockObserver.observers.pop();
  }

  localStorage.removeItem('jwttoken');
});

const mockUtility = {
  getBootstrap: () => {
    return {
      keySignatures: ['A', 'B', 'C', 'D', 'E', 'F', 'G'],
    };
  },
};

const renderWrapper = ({ children }) => {
  return (
    <BombayUtilityContext.Provider value={mockUtility}>
      {children}
    </BombayUtilityContext.Provider>
  );
};

const setUpModels = () => {
  const songNames = [
    "Hold On I'm Coming",
    'Walk This Way',
    'Kiss',
    'Cake By The Ocean',
    'Espresso',
    'Enter Sandman',
    'Stand By Me',
    "Ain't Talkin' 'Bout Love",
    'Kashmir',
    'War Pigs',
    'Renegade',
    'Refugee',
    'Gimme Shelter',
    'Bad Idea',
    'Smooth',
    'Baba O Riley',
    'Fireball',
    'Wear Your Love Like Heaven',
    'All Along The Watchtower',
    'Metal Guru',
  ];

  const artistNames = [
    'Same & Dave',
    'Aerosmith',
    'Prince',
    'DNCE',
    'Sabrina Carpenter',
    'Metallica',
    'Ben E. King',
    'Van Halen',
    'Led Zeppelin',
    'Black Sabbath',
    'Steppenwolf',
    'Tom Petty',
    'Rolling Stones',
    'Olivia Rodrigo',
    'Santana',
    'The Who',
    'Deep Purple',
    'Donovan',
    'Jimi Hendrix',
    'T. Rex',
  ];
  const [_fetchArtistBody, artistModels] = makeModels(
    10,
    {},
    'artist',
    (def) => {
      const idString = def.id || '1';
      const id = parseInt(idString) - 1;
      def.name = artistNames[id];
    },
  );

  const [_fetchBody, models] = makeModels(10, {}, 'song', (def) => {
    const idString = def.id || '1';
    const id = parseInt(idString) - 1;
    def.name = songNames[id];
    def.artist = artistModels[id];
  });

  const modelPromise = PromiseWithResolvers();
  mockModelFetcher.mockReturnValueOnce(modelPromise.promise);
  return [modelPromise, models];
};

it('Component should match snapshot', async () => {
  const [modelPromise, models] = setUpModels();

  const loginPromise = PromiseWithResolvers();
  mockLoginStatus.mockReturnValue(loginPromise);

  loginPromise.resolve({
    loggedIn: true,
    token: testToken,
  });

  const { asFragment } = render(
    <FakeContent>
      <SongList />
    </FakeContent>,
  );

  await act(async () => {
    modelPromise.resolve(models);
  });

  expect(asFragment()).toMatchSnapshot();
});

it('should render the list', async () => {
  const [modelPromise, models] = setUpModels();

  const loginPromise = PromiseWithResolvers();
  mockLoginStatus.mockReturnValue(loginPromise);
  loginPromise.resolve({
    loggedIn: true,
    token: testToken,
  });

  const result = render(
    <FakeContent>
      <SongList />
    </FakeContent>,
  );

  await act(async () => {
    modelPromise.resolve(models);
  });
  expect(mockObserver.mockObserver.observers.length).toBe(1);
  expect(screen.getByTestId('song-list-component')).toBeInTheDocument();
  expect(screen.getAllByTestId('song-list-card')).toHaveLength(models.length);
});

it('should render the next page', async () => {
  const [modelPromise1, models1] = setUpModels();
  const [modelPromise2, models2] = setUpModels();

  const loginPromise = PromiseWithResolvers();
  mockLoginStatus.mockReturnValue(loginPromise);
  loginPromise.resolve({
    loggedIn: true,
    token: testToken,
  });

  const result = render(
    <FakeContent>
      <SongList />
    </FakeContent>,
  );

  await act(async () => {
    modelPromise1.resolve(models1);
  });

  expect(mockObserver.mockObserver.observers.length).toBe(1);
  expect(screen.getByTestId('song-list-component')).toBeInTheDocument();
  expect(screen.getAllByTestId('song-list-card')).toHaveLength(models1.length);

  const observer = mockObserver.mockObserver.observers[0];

  await act(async () => {
    observer._fireIntersect(
      screen.getAllByTestId('song-list-card').slice(-1)[0],
    );
    modelPromise2.resolve(models2);
  });

  expect(screen.getAllByTestId('song-list-card')).toHaveLength(
    models1.length + models2.length,
  );
});

it('should stop when it runs out of data', async () => {
  const [modelPromise, models] = setUpModels();

  const loginPromise = PromiseWithResolvers();
  mockLoginStatus.mockReturnValue(loginPromise);
  loginPromise.resolve({
    loggedIn: true,
    token: testToken,
  });

  const result = render(
    <FakeContent>
      <SongList />
    </FakeContent>,
  );

  await act(async () => {
    modelPromise.resolve(models);
  });

  expect(mockObserver.mockObserver.observers.length).toBe(1);
  expect(screen.getByTestId('song-list-component')).toBeInTheDocument();
  expect(screen.getAllByTestId('song-list-card')).toHaveLength(models.length);

  const observer = mockObserver.mockObserver.observers[0];

  await act(async () => {
    observer._fireIntersect(
      screen.getAllByTestId('song-list-card').slice(-1)[0],
    );
    modelPromise.reject({ status: 404, message: 'Not Found' });
  });

  expect(mockObserver.mockObserver.observers.length).toBe(1);
  expect(screen.getByTestId('song-list-component')).toBeInTheDocument();
  expect(screen.getAllByTestId('song-list-card')).toHaveLength(models.length);
});

// Move this test to testing for the song editor form
it.skip('should add a song', async () => {
  // const loginPromise = PromiseWithResolvers();
  // mockLoginStatus.mockReturnValue(loginPromise);
  // loginPromise.resolve({
  //   loggedIn: true,
  //   token: testToken,
  // });
  // const modalRoot = document.getElementById("modal-root");
  // const result = render(
  //   <FakeContent>
  //     <SongList />
  //   </FakeContent>,
  // );
  // const [fetchBody] = makeModels(10, {}, "song");
  // const collectionUrl = Network.prepareURLFromArgs("song").toString();
  // await act(async () => {
  //   resolve(fetchBody);
  // });
  // expect(mockObserver.mockObserver.observers.length).toBe(1);
  // // const [, controls] = getAreas(result);
  // expect(modalRoot.childElementCount).toEqual(0);
  // // const addButton = controls.firstChild;
  // // act(() => {
  // //   addButton.click();
  // // });
  // expect(modalRoot.childElementCount).toEqual(1);
  // const submitButton = modalRoot.querySelector('[type="submit"]');
  // const [saveDef] = makeAModel("song");
  // const artistId = saveDef.artist.id;
  // await changeInput(
  //   modalRoot.querySelector('[data-targetfield="name"'),
  //   "input",
  //   "Herkimer",
  //   250,
  // );
  // await changeInput(
  //   modalRoot.querySelector('[data-targetfield="artist_id"'),
  //   "input",
  //   artistId,
  //   250,
  // );
  // await act(async () => {
  //   submitButton.click();
  // });
  // resolve(saveDef);
  // expect(Network.postToURLString).toBeCalledTimes(1);
  // expect(Network.postToURLString).toBeCalledWith(collectionUrl, {
  //   name: "Herkimer",
  //   artist_id: artistId,
  //   key_signature: "",
  //   tempo: null,
  //   lyrics: "",
  // });
});
