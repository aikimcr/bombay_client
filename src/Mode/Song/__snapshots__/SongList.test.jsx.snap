// Jest Snapshot v1, https://jestjs.io/docs/snapshot-testing

exports[`Component should match snapshot 1`] = `
<DocumentFragment>
  <div
    class="list-component"
    data-testid="song-list-component"
  >
    <div
      class="list-controls"
    >
      <button
        class="btn"
      >
        New
      </button>
      <div
        class="title"
      >
        Songs
      </div>
      <button
        class="btn"
      >
        Refresh
      </button>
    </div>
    <div
      class="song-list-container list-container"
    >
      <ul
        class="song-list card-list"
      >
        <li
          class="card"
          data-testid="song-list-card"
        >
          <div
            class="header"
          >
            Song
          </div>
          <div
            class="name"
          >
            Whats Up
          </div>
          <div
            class="details"
          >
            <div
              class="label"
            >
              Artist
            </div>
            <div
              class="text"
            >
              4 Non Blondes
            </div>
            <div
              class="label"
            >
              Key
            </div>
            <div
              class="text"
            >
              A
            </div>
            <div
              class="label"
            >
              Tempo
            </div>
            <div
              class="text"
            >
              60
            </div>
            <div
              class="label"
            >
              Lyrics
            </div>
            <div
              class="long-text"
            >
              <pre>
                Enim et qui esse aliquip aliqua minim cillum in enim consequat.
              </pre>
            </div>
          </div>
        </li>
        <li
          class="card"
          data-testid="song-list-card"
        >
          <div
            class="header"
          >
            Song
          </div>
          <div
            class="name"
          >
            Back In Black
          </div>
          <div
            class="details"
          >
            <div
              class="label"
            >
              Artist
            </div>
            <div
              class="text"
            >
              AC/DC
            </div>
            <div
              class="label"
            >
              Key
            </div>
            <div
              class="text"
            >
              A#
            </div>
            <div
              class="label"
            >
              Tempo
            </div>
            <div
              class="text"
            >
              80
            </div>
            <div
              class="label"
            >
              Lyrics
            </div>
            <div
              class="long-text"
            >
              <pre>
                Incididunt nulla enim dolore voluptate veniam est enim in ullamco magna et est ad.
              </pre>
            </div>
          </div>
        </li>
        <li
          class="card"
          data-testid="song-list-card"
        >
          <div
            class="header"
          >
            Song
          </div>
          <div
            class="name"
          >
            Chain of Fools
          </div>
          <div
            class="details"
          >
            <div
              class="label"
            >
              Artist
            </div>
            <div
              class="text"
            >
              Aretha Franklin
            </div>
            <div
              class="label"
            >
              Key
            </div>
            <div
              class="text"
            >
              Ab
            </div>
            <div
              class="label"
            >
              Tempo
            </div>
            <div
              class="text"
            >
              100
            </div>
            <div
              class="label"
            >
              Lyrics
            </div>
            <div
              class="long-text"
            >
              <pre>
                Sunt eu voluptate laboris exercitation sit non in dolore amet laborum Lorem mollit.
              </pre>
            </div>
          </div>
        </li>
        <li
          class="card"
          data-testid="song-list-card"
        >
          <div
            class="header"
          >
            Song
          </div>
          <div
            class="name"
          >
            Walk This Way
          </div>
          <div
            class="details"
          >
            <div
              class="label"
            >
              Artist
            </div>
            <div
              class="text"
            >
              Aerosmith
            </div>
            <div
              class="label"
            >
              Key
            </div>
            <div
              class="text"
            >
              Am
            </div>
            <div
              class="label"
            >
              Tempo
            </div>
            <div
              class="text"
            >
              120
            </div>
            <div
              class="label"
            >
              Lyrics
            </div>
            <div
              class="long-text"
            >
              <pre>
                Deserunt in tempor in cupidatat commodo quis aliquip ex id elit occaecat irure adipisicing est.
              </pre>
            </div>
          </div>
        </li>
        <li
          class="card"
          data-testid="song-list-card"
        >
          <div
            class="header"
          >
            Song
          </div>
          <div
            class="name"
          >
            Lets Stay Together
          </div>
          <div
            class="details"
          >
            <div
              class="label"
            >
              Artist
            </div>
            <div
              class="text"
            >
              Al Green
            </div>
            <div
              class="label"
            >
              Key
            </div>
            <div
              class="text"
            >
              A#m
            </div>
            <div
              class="label"
            >
              Tempo
            </div>
            <div
              class="text"
            >
              140
            </div>
            <div
              class="label"
            >
              Lyrics
            </div>
            <div
              class="long-text"
            >
              <pre>
                Velit eiusmod sunt reprehenderit dolor esse non Lorem fugiat enim adipisicing minim laboris.
              </pre>
            </div>
          </div>
        </li>
        <li
          class="card"
          data-testid="song-list-card"
        >
          <div
            class="header"
          >
            Song
          </div>
          <div
            class="name"
          >
            Hand In Pocket
          </div>
          <div
            class="details"
          >
            <div
              class="label"
            >
              Artist
            </div>
            <div
              class="text"
            >
              Alanis Morisette
            </div>
            <div
              class="label"
            >
              Key
            </div>
            <div
              class="text"
            >
              Abm
            </div>
            <div
              class="label"
            >
              Tempo
            </div>
            <div
              class="text"
            >
              160
            </div>
            <div
              class="label"
            >
              Lyrics
            </div>
            <div
              class="long-text"
            >
              <pre>
                Nostrud elit sunt et aliquip duis.
              </pre>
            </div>
          </div>
        </li>
        <li
          class="card"
          data-testid="song-list-card"
        >
          <div
            class="header"
          >
            Song
          </div>
          <div
            class="name"
          >
            The Thrill Is Gone
          </div>
          <div
            class="details"
          >
            <div
              class="label"
            >
              Artist
            </div>
            <div
              class="text"
            >
              B.B. King
            </div>
            <div
              class="label"
            >
              Key
            </div>
            <div
              class="text"
            >
              B
            </div>
            <div
              class="label"
            >
              Tempo
            </div>
            <div
              class="text"
            >
              60
            </div>
            <div
              class="label"
            >
              Lyrics
            </div>
            <div
              class="long-text"
            >
              <pre>
                Dolore et consectetur excepteur et culpa ad excepteur.
              </pre>
            </div>
          </div>
        </li>
        <li
          class="card"
          data-testid="song-list-card"
        >
          <div
            class="header"
          >
            Song
          </div>
          <div
            class="name"
          >
            Help!
          </div>
          <div
            class="details"
          >
            <div
              class="label"
            >
              Artist
            </div>
            <div
              class="text"
            >
              Beatles
            </div>
            <div
              class="label"
            >
              Key
            </div>
            <div
              class="text"
            >
              Bb
            </div>
            <div
              class="label"
            >
              Tempo
            </div>
            <div
              class="text"
            >
              80
            </div>
            <div
              class="label"
            >
              Lyrics
            </div>
            <div
              class="long-text"
            >
              <pre>
                In culpa dolore adipisicing magna aliqua.
              </pre>
            </div>
          </div>
        </li>
        <li
          class="card"
          data-testid="song-list-card"
        >
          <div
            class="header"
          >
            Song
          </div>
          <div
            class="name"
          >
            How Deep Is Your Love
          </div>
          <div
            class="details"
          >
            <div
              class="label"
            >
              Artist
            </div>
            <div
              class="text"
            >
              Bee Gees
            </div>
            <div
              class="label"
            >
              Key
            </div>
            <div
              class="text"
            >
              Bm
            </div>
            <div
              class="label"
            >
              Tempo
            </div>
            <div
              class="text"
            >
              100
            </div>
            <div
              class="label"
            >
              Lyrics
            </div>
            <div
              class="long-text"
            >
              <pre>
                Labore irure esse fugiat et veniam ex minim velit.
              </pre>
            </div>
          </div>
        </li>
        <li
          class="card"
          data-testid="song-list-card"
        >
          <div
            class="header"
          >
            Song
          </div>
          <div
            class="name"
          >
            Knock On Wood
          </div>
          <div
            class="details"
          >
            <div
              class="label"
            >
              Artist
            </div>
            <div
              class="text"
            >
              Ben E. King
            </div>
            <div
              class="label"
            >
              Key
            </div>
            <div
              class="text"
            >
              Bbm
            </div>
            <div
              class="label"
            >
              Tempo
            </div>
            <div
              class="text"
            >
              120
            </div>
            <div
              class="label"
            >
              Lyrics
            </div>
            <div
              class="long-text"
            >
              <pre>
                Et duis nulla cillum id labore nostrud mollit enim qui ut elit.
              </pre>
            </div>
          </div>
        </li>
      </ul>
    </div>
  </div>
</DocumentFragment>
`;
